cdc/entry/mounter.go

```diff
@@ -314,16 +383,94 @@ func datum2Column(tableInfo *model.TableInfo, datums map[int64]types.Datum, fill
 			// ApproximateBytes = column data size + column struct size
 			ApproximateBytes: colSize + sizeOfEmptyColumn,
 		}
+		columnIDs[tableInfo.RowColumnsOffset[colInfo.ID]] = colInfo.ID
+	}
+	return cols, rawCols, columnIDs, nil
+}
+
+// return error if cannot get the expected checksum from the decoder
+// return false if the checksum is not matched
+// return true if the checksum is matched and the checksum is the matched one.
+func (m *mounter) verifyChecksum(
+	columnIDs []int64, rawColumns []types.Datum, isPreRow bool,
+) (uint32, bool, error) {
+	if !m.integrity.Enabled() {
+		return 0, true, nil
+	}
+
+	// when the old value is not enabled, no previous columns so that no need to verify the checksum.
+	if isPreRow && !m.enableOldValue {
+		return 0, true, nil
 	}
-	return cols, rawCols, nil
+
+	var decoder *rowcodec.DatumMapDecoder
+	if isPreRow {
+		decoder = m.preDecoder
+	} else {
+		decoder = m.decoder
+	}
+	if decoder == nil {
+		return 0, false, errors.New("cannot found the decoder to get the checksum")
+	}
+
+	// if the checksum cannot be found, which means the upstream TiDB checksum is not enabled,
+	// so return matched as true to skip check the event.
+	first, ok := decoder.GetChecksum()
+	if !ok {
+		return 0, true, nil
+	}
+
+	checksum, err := m.encoder.Checksum(m.sctx, columnIDs, rawColumns)
+	if err != nil {
+		log.Error("failed to calculate the checksum", zap.Error(err))
+		return 0, false, errors.Trace(err)
+	}
+
+	// the first checksum matched, it hits in the most case.
+	if checksum == first {
+		return checksum, true, nil
+	}
+
+	extra, ok := decoder.GetExtraChecksum()
+	if !ok {
+		log.Error("cannot found the extra checksum, the first checksum mismatched",
+			zap.Uint32("checksum", checksum),
+			zap.Uint32("first", first),
+			zap.Uint32("extra", extra))
+		return checksum, false, errors.New("cannot found the extra checksum from the event")
+	}
+
+	if checksum == extra {
+		log.Warn("extra checksum matched, this may happen the upstream TiDB is during the DDL"+
```

Comment:
Please change it to debug level, or the log will make ticdc out of disk when there are many ddls in TiDB.

Reply:
Get it.